### PR TITLE
Add ability to overwrite the Kubernetes url and SSL parameters

### LINKED ISSUE #275: Kubernetes certificate verify failed
Hi
When I use fluented with k8s there are some errors, detail is below
```
error_class=Fluent::ConfigError error="Invalid Kubernetes API v1 endpoint https://172.26.199.1:443/api: SSL_connect returned=1 errno=0 state=error: certificate verify failed"
```
I believe it is because following config in ```kubernetes.conf```
```
@type kubernetes_metadata
```

According to this plugin, it can be configured with ```verify_ssl false```, but it has not environment variable for configuration. Could you add one configurable environment  variable for that? Or I make a pull request for that

### PR DESCRIPTION
**Pull request by aerostitch**
This does a little more than what was requested in #275 but I need to be able to use a custom url if I want to use my custom ca file so I figured I could add the `verify_ssl` at the same time.

Closes #275

### CODE REVIEW

File: templates/conf/kubernetes.conf.erb [hunk 1]
```diff
@@ -258,6 +258,9 @@
 <filter kubernetes.**>
   @type kubernetes_metadata
   @id filter_kube_metadata
+  kubernetes_url "#{ENV['KUBERNETES_URL'] || https://#{ENV['KUBERNETES_SERVICE_HOST']}:#{ENV['KUBERNETES_SERVICE_PORT']}/api}"
```
Comment (repeatedly) [inline templates/conf/kubernetes.conf.erb]: "KUBERNETES_" prefix seems general.
How about "FLUENTD_FILTER_KUBERNETS_URL" or something?
Other environment variable too.

Comment (aerostitch) [reply templates/conf/kubernetes.conf.erb]: For `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` this is referring to the standard variables that kubernetes makes available inside each pod but I guess you know that already.
I put them here to match the default https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter/blob/master/lib/fluent/plugin/filter_kubernetes_metadata.rb#L195 as we're specifying `kubernetes_url` here in any case.

I changed `KUBERNETES_URL` to `FLUENT_FILTER_KUBERNETS_URL`, rebased from master re-ran `make src-all`.

**Comment by idohalevi** [comment]
@repeatedly 
I need this too, when do you think you will merge?

File: docker-image/v1.3/debian-cloudwatch/conf/kubernetes.conf [hunk 2]
```diff
@@ -188,5 +188,8 @@
 <filter kubernetes.**>
   @type kubernetes_metadata
   @id filter_kube_metadata
+  kubernetes_url "#{ENV['FLUENT_FILTER_KUBERNETS_URL'] || https://#{ENV['KUBERNETES_SERVICE_HOST']}:#{ENV['KUBERNETES_SERVICE_PORT']}/api}"
```
Comment (ganmacs) [inline docker-image/v1.3/debian-cloudwatch/conf/kubernetes.conf]: I think this won't work.
How do you feel like the following code?
` "#{ENV['FLUENT_FILTER_KUBERNETS_URL'] || "https://#{ENV.fetch('KUBERNETES_SERVICE_HOST')}:#{ENV.fetch('KUBERNETES_SERVICE_PORT')}/api"}"`

File: docker-image/v1.3/debian-cloudwatch/conf/kubernetes.conf [hunk 3]
```diff
@@ -188,5 +188,8 @@
...
+  verify_ssl #{ENV['KUBERNETES_VERIFY_SSL'] || true}
```
[the remark this answers is not in the document]
Comment (aerostitch) [reply docker-image/v1.3/debian-cloudwatch/conf/kubernetes.conf]: Good catch, thanks!

File: docker-image/v1.3/debian-cloudwatch/conf/kubernetes.conf [hunk 4]
```diff
@@ -188,5 +188,8 @@
...
+  ca_file "#{ENV['KUBERNETES_CA_FILE'] || ''}"
```
[the remark this answers is not in the document]
Comment (aerostitch) [reply docker-image/v1.3/debian-cloudwatch/conf/kubernetes.conf]: Indeed, updated.

File: docker-image/v1.3/debian-cloudwatch/conf/kubernetes.conf (hunk 2 above)
Comment (aerostitch) [reply docker-image/v1.3/debian-cloudwatch/conf/kubernetes.conf]: That's fine with me too. Updated the PR.

**Comment by aerostitch** [comment]
Sorry for the late update I missed the latest notifications.
I updated the PR based on @repeatedly and @ganmacs 's feedbacks.
Let me know if you need me to do other changes.

Thanks!

**Comment by Timoses** [comment]
Why is this not in yet? It seems like a crucial part to get fluentd Kubernetes running on Clusters which use a custom CA for the Kubernetes API.

File: docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf [hunk 5]
```diff
@@ -188,5 +188,8 @@
 <filter kubernetes.**>
   @type kubernetes_metadata
   @id filter_kube_metadata
+  kubernetes_url "#{ENV['FLUENT_FILTER_KUBERNETS_URL'] || https://#{ENV['KUBERNETES_SERVICE_HOST']}:#{ENV['KUBERNETES_SERVICE_PORT']}/api}"
+  verify_ssl #{ENV['KUBERNETES_VERIFY_SSL'] || true}
```
Comment (ganmacs) [inline docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf]: https://github.com/fluent/fluentd-kubernetes-daemonset/pull/279/files/532befe36d84250a8d50b8cc8aeeaccd61206b1d#diff-bf350a335c264690ed81e269cca50f23

All others are the same.

File: docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf [hunk 6]
```diff
@@ -188,5 +188,8 @@
 <filter kubernetes.**>
   @type kubernetes_metadata
   @id filter_kube_metadata
+  kubernetes_url "#{ENV['FLUENT_FILTER_KUBERNETS_URL'] || https://#{ENV['KUBERNETES_SERVICE_HOST']}:#{ENV['KUBERNETES_SERVICE_PORT']}/api}"
```
Comment (ganmacs) [inline docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf]: https://github.com/fluent/fluentd-kubernetes-daemonset/pull/279/files/532befe36d84250a8d50b8cc8aeeaccd61206b1d#diff-bf350a335c264690ed81e269cca50f23

All others are the same.

File: docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf [hunk 7]
```diff
@@ -188,5 +188,8 @@
...
+  verify_ssl #{ENV['KUBERNETES_VERIFY_SSL'] || true}
+  ca_file "#{ENV['KUBERNETES_CA_FILE'] || ''}"
```
Comment (ganmacs) [inline docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf]: https://github.com/fluent/fluentd-kubernetes-daemonset/pull/279/files/532befe36d84250a8d50b8cc8aeeaccd61206b1d#diff-bf350a335c264690ed81e269cca50f23

All others are the same

**Comment by ganmacs** [comment]
Sorry for the delay. I added some comments.
And if possible, could you please write a more descriptive commit message... ("Update with the feedback from ganmacs" doesn't make sense to other developers... )

File: docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf (hunk 6 above)
Comment (aerostitch) [reply docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf]: Sorry for the late reply.
@ganmacs I don't understand your comment. Could you be more explicit in what change you are looking for please?

File: docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf (hunk 5 above)
Comment (aerostitch) [reply docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf]: @ganmacs I don't understand your comment. Could you be more explicit in what change you are looking for please?

File: docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf (hunk 7 above)
Comment (aerostitch) [reply docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf]: @ganmacs I don't understand your comment. Could you be more explicit in what change you are looking for please?

**Comment by aerostitch** [comment]
Sorry for the late reply.
There are some comments I really don't understand what you meant. Could you develop please?
For the commit message aren't you going to use "Squash and merge" when merging in master to avoid having several commits?

File: docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf (hunk 5 above)
Comment (aerostitch) [reply docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf]: Never mind, I finally understood! Sorry.

File: docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf (hunk 6 above)
Comment (aerostitch) [reply docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf]: Never mind, I finally understood! Sorry.

File: docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf (hunk 7 above)
Comment (aerostitch) [reply docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf]: Never mind, I finally understood! Sorry.

File: docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf (hunk 5 above)
Comment (aerostitch) [reply docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf]: Updated

**Comment by aerostitch** [comment]
Hi,

FYI I rebased the branch against the latest master, fixed a typo in a variable name, updated the template based on the remaining unaddressed feedback and regenerated the images for version 1.3, 1.4 and 1.7 (current) as this PR was addressing 1.3 when it was originally opened.
Let me know if there are still any blocker to get this merge please.

Thanks,
Joseph

**Comment by repeatedly** [comment]
Thanks for working on this.
I will release new image soon.

**Comment by aerostitch** [comment]
Awesome, thanks a lot! :)

**Pull request merged**